File: mobile_client/lib/services/team_service.dart
import 'package:collection/collection.dart';
import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/team.dart';
import 'package:mobile_client/constants/enum/team_creation_response.dart';
import 'package:mobile_client/constants/socket_events/room_event.dart';
import 'package:mobile_client/widgets/snackbar.dart';

import 'websocket_service.dart';

class TeamService extends ChangeNotifier {
  final WebSocketService _websocketService;
  List<Team> teams = [];
  String teamName = '';
  Team? team;

  TeamService(this._websocketService);

  Future<void> createTeam(String name, BuildContext context) async {
    teamName = name;
    final data = await _websocketService.sendWithAck<TeamId, String>(
        RoomEvent.createTeam.name, TeamId(name: name));
    final creationResponse = TeamCreationResponse.fromJson(data);

    if (!context.mounted) return;
    switch (creationResponse) {
      case TeamCreationResponse.teamNameTaken:
        showCustomSnackBar(
            message: "Nom de l'équipe existe déjà.", context: context);
        teamName = '';
        break;
      case TeamCreationResponse.maximumTeamLimitReached:
        showCustomSnackBar(
            message: "Le nombre maximal d'équipe a été atteint.",
            context: context);
        teamName = '';
        break;
      case TeamCreationResponse.teamCreated:
        break;
    }
  }

  void selectTeam(String name) {
    teamName = name;
    _websocketService.send(RoomEvent.selectTeam.name, TeamId(name: name));
  }

  Team? getTeam(String teamName) {
    return teams.firstWhereOrNull((team) => team.name == teamName);
  }

  void setup() {
    _websocketService.on<List<dynamic>>(RoomEvent.updateTeams.name, (data) {
      teams = data.map((team) => Team.fromJson(team)).toList();
      team = getTeam(teamName);
      notifyListeners();
    });
    _websocketService.send(RoomEvent.updateTeams.name);
  }

  void removeListeners() {
    _websocketService.removeAllListeners(RoomEvent.updateTeams.name);
  }

  void resetManager() {
    teams.clear();
    teamName = '';
    team = null;
    removeListeners();
    notifyListeners();
  }
}

File: desktop-client/src/app/game/services/team/team.service.ts
import { Injectable } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { TeamCreationResponse } from '@common/enum/team-creation-response';
import { Team, TeamId } from '@common/interfaces/team';
import { RoomEvent } from '@common/socket-events/room-event';
@Injectable({
    providedIn: 'root',
})
export class TeamService {
    teams: Team[] = [];
    teamName: string = '';
    team?: Team;

    constructor(
        private readonly websocketService: WebsocketService,
        private readonly snackBar: MatSnackBar,
    ) {}

    async createTeam(name: string) {
        this.teamName = name;
        const creationResponse = await this.websocketService.sendWithAck<TeamId, TeamCreationResponse>(RoomEvent.CREATE_TEAM, { name });
        switch (creationResponse) {
            case TeamCreationResponse.TEAM_NAME_TAKEN:
                this.snackBar.open("Nom de l'équipe existe déjà.");
                this.teamName = '';
                break;
            case TeamCreationResponse.MAXIMUM_TEAM_LIMIT_REACHED:
                this.snackBar.open("Le nombre maximal d'équipe a été atteint.");
                this.teamName = '';
                break;
            case TeamCreationResponse.TEAM_CREATED:
                break;
        }
    }

    selectTeam(name: string) {
        this.teamName = name;
        this.websocketService.send(RoomEvent.SELECT_TEAM, { name });
    }

    getTeam(teamName: string): Team | undefined {
        return this.teams.find((team) => team.name === teamName);
    }

    setup() {
        this.websocketService.on<Team[]>(RoomEvent.UPDATE_TEAMS, (teams?: Team[]) => {
            if (!teams) return;
            this.teams = teams;
            this.team = this.getTeam(this.teamName);
        });
        this.websocketService.send(RoomEvent.UPDATE_TEAMS);
    }

    removeListeners() {
        this.websocketService.removeAllListeners(RoomEvent.UPDATE_TEAMS);
    }

    resetManager() {
        this.teamName = '';
        this.team = undefined;
        this.teams = [];
        this.removeListeners();
    }
}
